jupyterhub/templates/hub/deployment.yaml

```diff
@@ -179,3 +179,12 @@ spec:
           ports:
             - containerPort: 8081
               name: hub
+          livenessProbe:
+            initialDelaySeconds: 30
```

Comment:
We set this delay because sometimes it can take the hub process a while to become ready correct? Does this also mean the minimum restart time/dead time is 30s (because the pod will only be considered live if the check succeeds which won't happen until at least 30s after the pod starts). This would be a penalty for small hubs that tend to restart quickly, there are probably many more small hubs than big hubs and small hubs are operated by less experienced operators. All this is a hunch, so we need to discuss if these assumptions are true. If we think they are true should we make it work for the small hub case by default with a note for experts on how to increase the delay?

One thing I'd check is the case where a hub would need (say) 60s to get ready but after 30s it gets restarted because the liveness probe fails. Now when the pod has restarted will it need another 60s to complete getting ready or has the time gotten shorter? If the time gets shorter then eventually the hub will start up within the 30s initial delay we give. If the startup time remains constant the hub will never manage to get live and will constantly be restarted.

If the latter case is true I think we should find a different way to determine liveness, maybe by checking if the hub process is running or something.

Another alternative would be that we remove the liveness probe because in almost all cases an unhealthy hub will crash and exit -> the pod will be restarted. Not sure if this is true but if it is this might be a nice way out of having to determine "how big should the initial delay be".

Reply:
Does keeping the delay short but increasing the `failureThreshold` work? https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/

Reply:
Does keeping the delay short but increasing the `failureThreshold` work? https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/

Reply:
> We set this delay because sometimes it can take the hub process a while to become ready correct?

Yepp!

> Does this also mean the minimum restart time/dead time is 30s (because the pod will only be considered live if the check succeeds which won't happen until at least 30s after the pod starts)

No the liveness probe adds another way to make the pod restart, but does not remove the ordinary way to crash and restart for that reason.

__About infinite restarts__
> One thing I'd check is the case where a hub would need (say) 60s to get ready but after 30s it gets restarted because the liveness probe fails. Now when the pod has restarted will it need another 60s to complete getting ready or has the time gotten shorter? If the time gets shorter then eventually the hub will start up within the 30s initial delay we give. If the startup time remains constant the hub will never manage to get live and will constantly be restarted.
>
>If the latter case is true I think we should find a different way to determine liveness, maybe by checking if the hub process is running or something.

I ended up noticing that there was a synchronous poll on previous user pods status, and that could fail and take 30 seconds to timeout. That caused liveness probe failures for me. This is why I reverted the original PR and now recreates it with this initial setup delay added. To be safe we could make it 60.

> Another alternative would be that we remove the liveness probe because in almost all cases an unhealthy hub will crash and exit -> the pod will be restarted.

With the current setup, we would crash on startups slower than 1 minute. If that happens, I would consider that a big failure, because the probe really match specifically to the hub container within the pod as compared to the full pod with potential init_containers etc.

Hmm... but... we have the hub upgrade situation, that should perhaps be allowed to take a while... Yeah I'm clueless on the time that may be required overall for that in worst case scenarios but I wouldn't want to mistakenly restart the hub pod during an upgrade.

I think due to the hub db upgrade complexity added, we should remove the livenessProbe of the hub, but keep the readinessProbe.

> (@manics) Does keeping the delay short but increasing the failureThreshold work? https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/

It would also work to increase the delay until restart, but it would restart slower if the livenessProbe started failing during normal operation, and we mostly wanted to increase it during normal operation.

Reply:
Thanks. For some reason I can never keep liveness and readyness probes straight in my head. "Liveness probe' determines if a pod needs to be restarted or not. "Readyness probe' determines if a pod is ready to serve traffic (a `Service` will only route traffic to a ready pod).

What I often need to look up is if a pod can be ready without the liveness probe having run. The answer is that a pod can be ready and receiving traffic if the liveness probe hasn't run yet.